### DISCUSSION TITLE
[공유] HTTP API vs REST API

### DISCUSSION THREAD

**Discussion by hyeonic** [woowacourse-study/2022-http-web-basic-for-all-developer#19] [General]
## API란?
API(Application Programming Interface)는 애플리케이션에서 사용할 수 있도록 운영체제나 프로그래밍 언어가 제공하는 기능을 제어할 수 있게 만든 인터페이스를 뜻한다. 정리하면 애플리케이션이 제공하는 기능을 사용할 수 있게 만든 매개체다. UI는 컴퓨터와 인간을 연결시키는 인터페이스를 의미한다. API는 컴퓨터나 소프트웨어를 서로 연결하는데 사용한다.

## HTTP API란?

HTTP를 사용하여 프로그램끼리 소통하는 API를 말한다. 보통 흔히 보는 OPEN API 등의 대부분 API는 HTTP 라는 통신 규약으로 소통하는 API이다.

## REST API란?

REST(Representational State Transfer)란 네트워크 아키텍처 스타일이다. 여기서 `네트워크 아키텍처 스타일`란 네트워크 자원을 정의하고 처리하는 방법 전반을 일컫는다.

2000년도에 로이 필딩 (Roy Fielding)의 박사학위 논문에서 최초로 소개되었다. 로이 필딩은 HTTP의 주요 저자 중 한 사람으로 그 당시 웹(HTTP) 설계의 우수성에 비해 제대로 사용하지 못하는 모습에 안타까워하며 HTTP의 장점을 최대한 활용할 수 있는 아키텍처로써 REST를 발표했다.

즉, REST는 HTTP를 잘 활용하기 위한 원칙이라고 할 수 있고, REST API는 이 원칙을 준수해 만든 API이다.

## 강의 답변 중 일부

강의 중 질문의 답변에 따르면, HTTP API와 REST API는 사실 거의 같은 의미로 사용하고 있다고 한다. 정확한 차이점으로 `REST API`는 HTTP 프로토콜을 따르면서 아래의 4가지 가이드 원칙을 지켜야 한다.

1) 자원의 식별
2) 메세지를 통한 리소스 조작
3) 자기서술적 메세지
4) 애플리케이션의 상태에 대한 엔진으로서 하이퍼미디어(HATEOAS)

이러한 제약 조건들을 완벽하게 지키면서 개발하는 것을 `RESTful API`라고 한다. 하지만 실무에서는 이런 방법으로 개발하는 것은 현실적으로 어렵고 개발비용 대비 효과가 크지 않다. 특히 4번째 원칙이 구현하기 어렵다고 한다. 그런데 이미 많은 사람들이 위 조건들을 지키지 않아도 REST API라고 하기 때문에 HTTP API와 같은 의미로 사용하고 있다고 한다. 하지만 비로소 위 제약 조건들을 모두 지켜야 REST API라고 말할 수 있다.

## References

* [API, HTTP API, REST API 차이](https://bentist.tistory.com/37)
* [HTTP API vs REST API](https://www.inflearn.com/questions/126743)
* [REST API에 대한 질문이 있습니다](https://www.inflearn.com/questions/152610)

**Comment by kth990303** [comment]
너무 Restful 하게 작성하려고 고민하는 데에도 비용이 꽤 있다고 하네요! 소통에 크게 문제되지 않는다면 융통성있게 처리해도 괜찮을 것 같다는 개인적인 생각입니다 :) 좋은 글 잘봤습니다! 👍
